lib/spack/spack/compilers/msvc.py

```diff
@@ -198,11 +191,34 @@ def __init__(self, *args, **kwargs):
         # paths[2] refers to the fc path and is a generic check
         # for a fortran compiler
         if paths[2]:
+
+            def get_oneapi_root(pth: str):
+                """From within a prefix known to be a oneAPI path
+                determine the oneAPI root path from arbitrary point
+                under root
+
+                Args:
+                    pth: path prefixed within oneAPI root
+                """
+                if not pth or not os.path.basename(pth):
+                    return ""
+                if os.path.basename(pth) == "oneAPI":
+                    return pth
+                return get_oneapi_root(os.path.dirname(pth))
+
             # If this found, it sets all the vars
-            oneapi_root = os.path.join(self.cc, "../../..")
+            oneapi_root = get_oneapi_root(self.fc)
+            if not oneapi_root:
+                raise RuntimeError(f"Non oneAPI Fortran compiler {self.fc} assigned to MSVC")
             oneapi_root_setvars = os.path.join(oneapi_root, "setvars.bat")
+            # some oneAPI versions return a version more precise than their
+            # install paths
+            version_from_path_grp = re.search(r"([1-9][0-9]*\.[0-9]*(?:\.[0-9])*)", self.fc)
+            version_from_path = (
+                "latest" if not version_from_path_grp else version_from_path_grp.group(1)
+            )
```

Comment:
I'd recommend the following regex:

`(\d+\.\d+(?:\.\d+)?`

(to clarify, not only did I use `\d`, but the 2nd/3rd components are 1-or-more, vs. 0-or-more), and I think you wanted a `?` for the last pattern to accommodate both `1.2` and `1.2.3`)

I'd also recommend folding in `latest`, because that may not in fact be whatever the `fc` path is, so like

```
numver = r"\d+\.\d+(?:\.\d+)?"
pattern = f"((?:{numver})|(?:latest))"
```

Reply:
Looks great, thanks!